p2p/transport/webrtc/listener.go

```diff
@@ -0,0 +1,350 @@
+				defer cancelFunc()
+				conn, err := l.accept(ctx, addr)
+				if err != nil {
+					log.Debugf("could not accept connection: %v", err)
+					return
+				}
+				l.connChan <- conn
+			}()
+		}
+	}
+}
+
+func (l *listener) Accept() (tpt.CapableConn, error) {
+	select {
+	case <-l.closeChan:
+		return nil, os.ErrClosed
+	case conn := <-l.connChan:
+		return conn, nil
+	}
+}
+
+func (l *listener) Close() error {
+	select {
+	case <-l.closeChan:
+		return nil
+	default:
+	}
+	close(l.closeChan)
+	return nil
+}
+
+func (l *listener) Addr() net.Addr {
+	return l.mux.LocalAddr()
+}
+
+func (l *listener) Multiaddr() ma.Multiaddr {
+	return l.localMultiaddr
+}
+
+func (l *listener) accept(ctx context.Context, addr candidateAddr) (tpt.CapableConn, error) {
+	localFingerprint := strings.ReplaceAll(l.localFingerprint.Value, ":", "")
+	// signaling channel
+	signalChan := make(chan struct{})
+
+	se := webrtc.SettingEngine{}
+	se.SetAnsweringDTLSRole(webrtc.DTLSRoleServer)
+	se.DetachDataChannels()
+	se.DisableCertificateFingerprintVerification(true)
+	se.SetICECredentials(addr.ufrag, localFingerprint)
+	se.SetLite(true)
+	se.SetICEUDPMux(l.mux)
+
+	api := webrtc.NewAPI(webrtc.WithSettingEngine(se))
+
+	clientSdpString := renderClientSdp(sdpArgs{
+		Addr:        addr.raddr,
+		Fingerprint: defaultMultihash,
+		Ufrag:       addr.ufrag,
+		Password:    addr.ufrag,
+	})
+	clientSdp := webrtc.SessionDescription{SDP: clientSdpString, Type: webrtc.SDPTypeOffer}
+	pc, err := api.NewPeerConnection(l.config)
+	if err != nil {
+		return nil, err
+	}
+
+	var connectedOnce sync.Once
+	pc.OnConnectionStateChange(func(state webrtc.PeerConnectionState) {
+		if state == webrtc.PeerConnectionStateConnected {
+			connectedOnce.Do(func() { signalChan <- struct{}{} })
+		}
+	})
+
+	pc.SetRemoteDescription(clientSdp)
+
+	answer, err := pc.CreateAnswer(nil)
+	if err != nil {
+		_ = pc.Close()
+		return nil, err
+	}
+	pc.SetLocalDescription(answer)
+
+	// await peerconnection connected state
+	select {
+	case <-signalChan:
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	}
+
+	// await openening of datachannel
+	dcChan := make(chan *DataChannel)
+	// this enforces that the correct data channel label is used
+	// for the handshake
+	handshakeChannel, err := pc.CreateDataChannel("data", &webrtc.DataChannelInit{
+		Negotiated: func(v bool) *bool { return &v }(true),
+		ID:         func(v uint16) *uint16 { return &v }(1),
+	})
+	handshakeChannel.OnOpen(func() {
+		detached, err := handshakeChannel.Detach()
+		if err != nil {
+			return
+		}
+		dcChan <- newDataChannel(
+			detached,
+			pc,
+			l.mux.LocalAddr(),
+			addr.raddr,
+		)
+	})
+
+	timeout := 10 * time.Second
+	var dc *DataChannel = nil
+	select {
+	case <-time.After(timeout):
+		_ = pc.Close()
+		return nil, ErrDataChannelTimeout
+	case dc = <-dcChan:
+		if dc == nil {
+			_ = pc.Close()
+			return nil, fmt.Errorf("should be unreachable")
+		}
+		// clear to ignore opening of future data channels
```

Comment:
What is getting cleared here? Did we set any `OnDataChannel` before? If so, this would be racy, wouldn't it?

Reply:
The callback is internally protected by a mutex: https://github.com/pion/webrtc/blob/master/peerconnection.go#L278 . I've also wrapped the data channel detachment logic in a `sync.Once`. At this point, we do not care if any other datachannels are opened until the noise handshake is complete.